src/com/xilinx/rapidwright/rwroute/RouterHelper.java

```diff
@@ -371,28 +373,56 @@ private static boolean isInvertibleDSPBELPin(BELPin belPin) {
         Net staticNet = design.getGndNet();
         Set<SitePinInst> toInvertPins = new HashSet<>();
-        for (SitePinInst currSitePinInst : pins) {
+        nextSitePin: for (SitePinInst currSitePinInst : pins) {
             if (!currSitePinInst.getNet().equals(staticNet))
                 throw new RuntimeException(currSitePinInst.toString());
-            BELPin[] belPins = currSitePinInst.getSiteInst().getSiteWirePins(currSitePinInst.getName());
-            // DSP or BRAM
-            if (belPins.length == 2) {
-                for (BELPin belPin : belPins) {
-                    if (belPin.isSitePort())    continue;
-                    if (currSitePinInst.getSite().getName().startsWith("RAM")) {
-                        if (!belPin.getBEL().canInvert()) {
-                            continue;
-                        }
-                        if (belPin.getBELName().startsWith("CLK")) {
-                            continue;
-                        }
-                        toInvertPins.add(currSitePinInst);
-                    } else if (currSitePinInst.getSite().getName().startsWith("DSP")) {
-                        if (isInvertibleDSPBELPin(belPin)) {
+            if (currSitePinInst.isLUTInputPin()) {
+                Collection<Cell> connectedCells = DesignTools.getConnectedCells(currSitePinInst);
+                for (Cell cell : connectedCells) {
+                    if (!LUTTools.isCellALUT(cell)) {
+                        continue nextSitePin;
+                    }
+                }
+
+                toInvertPins.add(currSitePinInst);
+
+                for (Cell cell : connectedCells) {
+                    // Find the logical pin name
+                    String physicalPinName = "A" + currSitePinInst.getName().charAt(1);
+                    String logicalPinName = cell.getLogicalPinMapping(physicalPinName);
+
+                    // Get the LUT equation
+                    String lutEquation = LUTTools.getLUTEquation(cell);
+                    assert(lutEquation.contains(logicalPinName));
+
+                    // Compute a new LUT equation with that logical input inverted
+                    String newLutEquation = lutEquation.replace(logicalPinName, "!" + logicalPinName)
+                            // Cancel out double inversions
+                            .replace("!!", "");
+                    // TODO: This may modify an folded EDIFCellInst!!!
+                    LUTTools.configureLUT(cell, newLutEquation);
+                }
+            } else {
+                BELPin[] belPins = currSitePinInst.getSiteInst().getSiteWirePins(currSitePinInst.getName());
+                // DSP or BRAM
+                if (belPins.length == 2) {
+                    for (BELPin belPin : belPins) {
+                        if (belPin.isSitePort()) continue;
+                        if (currSitePinInst.getSite().getName().startsWith("RAM")) {
```

Comment:
URAMs also can have invertible inputs:
![image](https://github.com/Xilinx/RapidWright/assets/20803522/b8bcfd06-b3a6-44e2-9474-952d2ec9a4e6)

Reply:
You're absolutely right. This PR just indents what we had before here, but I'll add a note that those BELs (plus others) are other things that can be identified in the future.